### PR TITLE
Fixed scaling script Bugs (Bug Numbers:22165642,22074935,21841360,222…

### PR DESCRIPTION
**Pull request by ptgadmin-cse**
…37991)

### CODE REVIEW

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 1]
```diff
@@ -355,19 +345,28 @@ if ($hostpoolInfo.LoadBalancerType -eq "DepthFirst") {
                     if ($sessionHost.Sessions -ge $SessionhostLimit) {
                         foreach ($sHost in $getHosts) {
                             if ($sHost.Status -eq "Available" -and $sHost.Sessions -eq 0) { break }
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: What happens if session host is "Available" but has more than "0" sessions?

File: wvd-sh/WVD scaling script/Config-MSI.Json [hunk 2]
```diff
@@ -29,22 +29,27 @@
...
 					"Value": "BreadthFirst"
 				},
 				{
 					"_Comments": "Session Load Balancing Peak Hours",
 					"Name": "sessionLoadBalancingPeakHours",
 					"Value": "9:00-3:00"
 				},
+				{
+					"_Comments": "Name of the Maintainance Tag",
+					"Name": "maintainanceTagName",
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/Config-MSI.Json]: Replace "maintananceTagName" with "maintanenceTagName" everywhere.

File: wvd-sh/WVD scaling script/Config-MSI.Json [hunk 3]
```diff
@@ -29,22 +29,27 @@
...
+					"Value": "maintainance"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/Config-MSI.Json]: Replace "maintanance" with "maintanence" everywhere.

File: wvd-sh/WVD scaling script/Config.json [hunk 4]
```diff
@@ -40,31 +39,36 @@
...
 					"Value": "DepthFirst"
 				},
 				{
 					"_Comments": "Session Load Balancing Peak Hours",
 					"Name": "sessionLoadBalancingPeakHours",
 					"Value": "9:00-18:00"
+				},
+				{
+					"_Comments": "Name of the Maintainance Tag",
+					"Name": "maintainanceTagName",
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/Config.json]: Replace "maintananceTagName" with "maintanenceTagName" everywhere.

File: wvd-sh/WVD scaling script/Config.json [hunk 5]
```diff
@@ -40,31 +39,36 @@
...
+					"Value": "maintainance"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/Config.json]: Replace "maintanance" with "maintanence" everywhere.

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 6]
```diff
@@ -312,18 +293,27 @@ if ($hostpoolInfo.LoadBalancerType -eq "DepthFirst") {
...
-                        if ($sessionhost.Status -eq "NoHeartbeat") {
-                            $sessionhostname = $sessionhost.sessionhostname
+                        #Check the session host status and session host is healthy or not before starting the host
+                        if ($sessionhost.Status -eq "NoHeartbeat" -and $sessionhost.UpdateState -eq "Succeeded") {
+                            $sessionhostname = $sessionhost.sessionhostname | out-string
+                                $VMName = $sessionhostname.Split(".")[0]
+                                $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                                #Check the Session host is in Maintainance
+                                if($VmInfo.Tags.Keys -contains  $maintainanceTagName){
+                                Write-Log 1 "Session Host is in Maintainace: $sessionhostname"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 7]
```diff
@@ -497,27 +498,40 @@ if ($hostpoolInfo.LoadBalancerType -eq "DepthFirst") {
                                 }
                             }
                         }
+
+
                         $VMName = $sessionHost.Split(".")[0]
+                          #Check the Session host is in maintainance
+                                $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                                if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+                                Write-Log 1 "Session Host is in Maintainace: $($sessionHost | out-string)"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 8]
```diff
@@ -571,24 +585,32 @@ if ($hostpoolInfo.LoadBalancerType -eq "DepthFirst") {
...
+                            if ($sessionHost.Status -eq "NoHeartbeat" -and $sessionhost.UpdateState -eq "Succeeded") {
                                 Write-Log 1 "Existing Sessionhost Sessions value reached near by hostpool maximumsession limit need to start the session host" "Info"
-                                $sessionhostname = $sessionHost.sessionhostname
+                                $sessionhostname = $sessionHost.sessionhostname | Out-String
+                                
+                                $VMName = $sessionHostname.Split(".")[0]
+                                $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                                #Check the Session host is in Maintainance
+                                if($VmInfo.Tags.Keys -eq $maintainanceTagName){
+                                Write-Log 1 "Session Host is in Maintainace: $sessionhostname"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 9]
```diff
@@ -691,16 +719,26 @@ else {
                 #check whether the number of running VMs meets the minimum or not
                 if ($numberOfRunningHost -lt $MinimumNumberOfRDSH) {
                     $VMName = $sessionHost.Split(".")[0]
+                     $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                    #Check the Session host is in Maintainance
+                    if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+                        Write-Log 1 "Session Host is in Maintainace: $($sessionhost | out-string )"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 10]
```diff
@@ -742,19 +780,26 @@ else {
                 foreach ($sessionHost in $RDSessionHost.sessionhostname) {
                     if ($hostPoolUserSessions.Count -ge $AvailableSessionCapacity) {
                         $VMName = $sessionHost.Split(".")[0]
+                          $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                            #Check the Session host is in Maintainance
+                            if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+                                 Write-Log 1 "Session Host is in Maintainace: $($sessionhost | out-string)"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 11]
```diff
@@ -742,19 +780,26 @@ else {
...
+                                }
+
                         $roleInstance = Get-AzureRmVM -Status | Where-Object { $_.Name.Contains($VMName) }
 
                         if ($sessionHost.ToLower().Contains($roleInstance.Name.ToLower())) {
-                            #check if the Azure VM is running or not
-
-                            if ($roleInstance.PowerState -ne "VM running") {
-                                $getShsinfo = Get-RdsSessionHost -TenantName $tenantname -HostPoolName $hostPoolName
+                            #check if the Azure VM is running or not and session host is healthy or not
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: "check if the Azure VM is running and if the session host is healthy"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 12]
```diff
@@ -742,19 +780,26 @@ else {
...
+                            $getShsinfo = Get-RdsSessionHost -TenantName $tenantname -HostPoolName $hostPoolName -Name $sessionHost
+                            if ($roleInstance.PowerState -ne "VM running" -and $getShsinfo.UpdateState -eq "Succeeded") {
+                                #Check the session host is allowing new sessions or not
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: "check if the session host is allowing new connections"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 13]
```diff
@@ -667,12 +689,18 @@ else {
...
-
-            $VMName = $sessionHost.Split(".")[0]
+        foreach ($sessionHost in $RDSessionHost) {
+            Write-Log 1 "Checking session host:$($sessionHost.SessionHostName | Out-String)  of sessions:$($sessionHost.Sessions) and status:$($sessionHost.Status)" "Info"
+            $hostName = $sessionHost.SessionHostName | Out-String
+            $VMName = $hostName.Split(".")[0]
+            $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+             #Check the Session host is in Maintainance
+             if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+             Write-Log 1 "Session Host is in Maintainace: $hostName"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 14]
```diff
@@ -951,9 +988,20 @@ else {
...
                                 }
+                                
+                              
                                 #check the session count before shutting down the VM
                                 if ($existingSession -eq 0) {
 
+                                    #Check the Session host is in maintainance
+                                    $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                                    if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+                                    Write-Log 1 "Session Host is in Maintainace: $($sessionHost | out-string)"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 [hunk 15]
```diff
@@ -1036,35 +1087,38 @@ else {
...
-                    $sessionHost = $sessionHosts.SessionHostname
-                    $VMName = $sessionHost.Split(".")[0]
+                foreach ($sessionHost in $RDSessionHost) {
+                    $hostname = $sessionHost.SessionHostname | out-string
+                    $VMName = $hostname.Split(".")[0]
+                    
+                    $VmInfo = Get-AzureRmVM | Where-Object { $_.Name -eq $VMName }
+                    #Check the Session host is in Maintainance
+                    if($VmInfo.Tags.Keys -contains $maintainanceTagName){
+                        Write-Log 1 "Session Host is in Maintainace: $hostname"
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Replace "Maintainace" with "maintenance"

File: wvd-sh/WVD scaling script/basicScale.ps1 [hunk 16]
```diff
@@ -305,17 +294,27 @@ if ($hostpoolInfo.LoadBalancerType -eq "DepthFirst") {
                 if ($numberOfRunningHost -lt $MinimumNumberOfRDSH) {
                     $hostsessions = $sessionHost.Sessions
                     if ($hostpoolMaxSessionLimit -ne $hostsessions) {
-                        if ($sessionhost.Status -eq "NoHeartbeat") {
-                            $sessionhostname = $sessionhost.sessionhostname
+						#Check the session host status and session host is healthy or not before starting the host
```
Comment (ChristianMontoya) [inline wvd-sh/WVD scaling script/basicScale.ps1]: All of the same changes from "basicScale-MSI.ps1" applies to "basicScale.ps1"

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 6 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: We have replaced

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 7 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: We have replaced in both scripts

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 8 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: We have replaced in both scripts

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 9 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: We have replaced in both scripts

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 10 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: We have replaced in both scripts

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 1 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Thank you Christian, we have updated accordingly

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 11 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Thank you Christian, we have updated accordingly

File: wvd-sh/WVD scaling script/basicScale.ps1 (hunk 16 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale.ps1]: Thank you Christian for to addressing us. We have checked and updated in both scripts

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 12 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: Thank you Christian, we have updated accordingly

File: wvd-sh/WVD scaling script/basicScale-MSI.ps1 (hunk 1 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/basicScale-MSI.ps1]: If reached the more than "0" sessions script will automatically spin up the one more VM.

File: wvd-sh/WVD scaling script/Config-MSI.Json (hunk 2 above)
Comment (ptgadmin-cse) [reply wvd-sh/WVD scaling script/Config-MSI.Json]: Thank you Christian to addressing us. We have replaced in both the scripts.

**Comment by ChristianMontoya** [comment]
Talked to Rahul. Changes and clearer logic coming up in edits for next week. Approving since continues to work and addresses bugs.

**Pull request merged**